### PR TITLE
Fixed a memory corruption issue with JS memory operations that were r…

### PR DESCRIPTION
**Pull request by john-sharratt**
Fixed a memory corruption issue with JS memory operations that were reusing invalid views

### CODE REVIEW

File: examples/exports_function.rs [hunk 1]
```diff
@@ -40,7 +40,7 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
     // the default provided by Wasmer.
     // You can use `Store::default()` for that.
     let mut store = Store::new(Cranelift::default());
-    let mut env = FunctionEnv::new(&mut store, ());
+    let _env = FunctionEnv::new(&mut store, ());
```
Comment (syrusakbary) [inline examples/exports_function.rs:43]: If it's not used, we can just remove this line :)

File: examples/errors.rs [hunk 2]
```diff
@@ -39,7 +39,7 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
     // the default provided by Wasmer.
     // You can use `Store::default()` for that.
     let mut store = Store::new(Cranelift::default());
-    let mut env = FunctionEnv::new(&mut store, ());
+    let _env = FunctionEnv::new(&mut store, ());
```
Comment (syrusakbary) [inline examples/errors.rs:42]: If it's not used, we can just remove this line :)

File: examples/engine_headless.rs [hunk 3]
```diff
@@ -96,7 +96,7 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
         // We create a headless Universal engine.
         let engine = EngineBuilder::headless();
         let mut store = Store::new(engine);
-        let mut env = FunctionEnv::new(&mut store, ());
+        let _env = FunctionEnv::new(&mut store, ());
```
Comment (syrusakbary) [inline examples/engine_headless.rs:99]: If it's not used, we can just remove this line :)

File: examples/exports_memory.rs [hunk 4]
```diff
@@ -37,7 +37,7 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
     // the default provided by Wasmer.
     // You can use `Store::default()` for that.
     let mut store = Store::new(Cranelift::default());
-    let mut env = FunctionEnv::new(&mut store, ());
+    let _env = FunctionEnv::new(&mut store, ());
```
Comment (syrusakbary) [inline examples/exports_memory.rs:40]: If it's not used, we can just remove this line :)

File: examples/features.rs [hunk 5]
```diff
@@ -41,7 +41,7 @@ fn main() -> anyhow::Result<()> {
 
     // Now, let's define the store, and compile the module.
     let mut store = Store::new(engine);
-    let mut env = FunctionEnv::new(&mut store, ());
+    let _env = FunctionEnv::new(&mut store, ());
```
Comment (syrusakbary) [inline examples/features.rs:44]: If it's not used, we can just remove this line :)

File: examples/instance.rs [hunk 6]
```diff
@@ -39,7 +39,7 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
     // the default provided by Wasmer.
     // You can use `Store::default()` for that.
     let mut store = Store::new(Cranelift::default());
-    let mut env = FunctionEnv::new(&mut store, ());
+    let _env = FunctionEnv::new(&mut store, ());
```
Comment (syrusakbary) [inline examples/instance.rs:42]: If it's not used, we can just remove this line :)

File: examples/memory.rs [hunk 7]
```diff
@@ -59,7 +59,7 @@ fn main() -> anyhow::Result<()> {
     // the default provided by Wasmer.
     // You can use `Store::default()` for that.
     let mut store = Store::new(Cranelift::default());
-    let mut env = FunctionEnv::new(&mut store, ());
+    let _env = FunctionEnv::new(&mut store, ());
```
Comment (syrusakbary) [inline examples/memory.rs:62]: If it's not used, we can just remove this line :)

File: examples/metering.rs [hunk 8]
```diff
@@ -72,7 +72,7 @@ fn main() -> anyhow::Result<()> {
     // We use our previously create compiler configuration
     // with the Universal engine.
     let mut store = Store::new(EngineBuilder::new(compiler_config));
-    let mut env = FunctionEnv::new(&mut store, ());
+    let _env = FunctionEnv::new(&mut store, ());
```
Comment (syrusakbary) [inline examples/metering.rs:75]: If it's not used, we can just remove this line :)

File: examples/tunables_limit_memory.rs [hunk 9]
```diff
@@ -143,7 +143,7 @@ fn main() -> Result<(), Box<dyn std::error::Error>> {
 
     // Create a store, that holds the engine and our custom tunables
     let mut store = Store::new_with_tunables(compiler, tunables);
-    let mut env = FunctionEnv::new(&mut store, ());
+    let _env = FunctionEnv::new(&mut store, ());
```
Comment (syrusakbary) [inline examples/tunables_limit_memory.rs:146]: If it's not used, we can just remove this line :)

File: lib/api/tests/js_externals.rs [hunk 10]
```diff
@@ -46,7 +46,7 @@ mod js {
     #[wasm_bindgen_test]
     fn global_set() {
         let mut store = Store::default();
-        let mut env = FunctionEnv::new(&mut store, ());
+        let _env = FunctionEnv::new(&mut store, ());
```
Comment (syrusakbary) [inline lib/api/tests/js_externals.rs:49]: If it's not used, we can just remove this line :)

File: lib/api/tests/js_externals.rs [hunk 11]
```diff
@@ -29,7 +29,7 @@ mod js {
     #[wasm_bindgen_test]
     fn global_get() {
         let mut store = Store::default();
-        let mut env = FunctionEnv::new(&mut store, ());
+        let _env = FunctionEnv::new(&mut store, ());
```
Comment (syrusakbary) [inline lib/api/tests/js_externals.rs:32]: If it's not used, we can just remove this line :)

File: lib/api/tests/js_externals.rs [hunk 12]
```diff
@@ -6,7 +6,7 @@ mod js {
     #[wasm_bindgen_test]
     fn global_new() {
         let mut store = Store::default();
-        let mut env = FunctionEnv::new(&mut store, ());
+        let _env = FunctionEnv::new(&mut store, ());
```
Comment (syrusakbary) [inline lib/api/tests/js_externals.rs:9]: If it's not used, we can just remove this line :)

File: lib/api/src/js/externals/memory_view.rs [hunk 13]
```diff
@@ -0,0 +1,186 @@
...
+    /// ```
+    pub fn size(&self) -> Pages {
+        Bytes(self.size as usize).try_into().unwrap()
+    }
+
+    /// Used by tests
+    #[doc(hidden)]
+    pub fn uint8view(&self) -> js_sys::Uint8Array {
+        self.view.clone()
+    }
```
Comment (syrusakbary) [inline lib/api/src/js/externals/memory_view.rs:86]: Now that we have the `MemoryView` I believe we can remove all usages of this function `uint8view`, and make them use the read, write instead. That way the API is fully uniform between js/sys

File: lib/api/Cargo.toml [hunk 14]
```diff
@@ -26,6 +26,7 @@ indexmap = { version = "1.6", features = ["serde-1"] }
 cfg-if = "1.0"
 thiserror = "1.0"
 more-asserts = "0.2"
+tracing = "0.1"
```
Comment (syrusakbary) [inline lib/api/Cargo.toml:29]: can we have this as an optional dependency depending on a feature flag?

File: lib/api/tests/js_instance.rs [hunk 15]
```diff
@@ -511,12 +511,12 @@ mod js {
             args: &[Val],
         ) -> Result<Vec<Val>, RuntimeError> {
             let memory: &Memory = env.data().memory.as_ref().unwrap();
-            let memory_val = memory.uint8view(&env).get_index(0);
+            let memory_val = memory.view(&env).uint8view().get_index(0);
```
Comment (syrusakbary) [inline lib/api/tests/js_instance.rs:514]: Let's try to use the MemoryView API directly

File: lib/api/tests/js_instance.rs [hunk 16]
```diff
@@ -535,12 +535,12 @@ mod js {
...
         let memory = instance.exports.get_memory("memory").unwrap();
-        assert_eq!(memory.data_size(&store), 65536);
-        let memory_val = memory.uint8view(&store).get_index(0);
+        assert_eq!(memory.view(&store).data_size(), 65536);
+        let memory_val = memory.view(&store).uint8view().get_index(0);
         assert_eq!(memory_val, 0);
 
-        memory.uint8view(&store).set_index(0, 2);
-        let memory_val = memory.uint8view(&store).get_index(0);
+        memory.view(&store).uint8view().set_index(0, 2);
```
Comment (syrusakbary) [inline lib/api/tests/js_instance.rs:542]: Let's try to use the MemoryView API directly, so we can remove the `uint8view` dependency

File: lib/api/tests/js_instance.rs [hunk 17]
```diff
@@ -552,7 +552,7 @@ mod js {
         assert_eq!(exported.call(&mut store, &[Val::I32(4)]), Ok(expected));
 
         // It works if we update the memory
-        memory.uint8view(&store).set_index(0, 3);
+        memory.view(&store).uint8view().set_index(0, 3);
```
Comment (syrusakbary) [inline lib/api/tests/js_instance.rs:555]: Let's try to use the MemoryView API directly, so we can remove the `uint8view` dependency

File: lib/api/src/sys/externals/memory.rs [hunk 18]
```diff
@@ -73,61 +76,10 @@ impl Memory {
...
-    /// #
-    /// let m = Memory::new(&mut store, MemoryType::new(1, None, false)).unwrap();
-    ///
-    /// assert_eq!(m.size(&mut store), Pages(1));
-    /// ```
-    pub fn size(&self, store: &impl AsStoreRef) -> Pages {
-        self.handle.get(store.as_store_ref().objects()).size()
+    /// Creates a view into the memory that then allows for
+    /// read and write 
+    pub fn view<'a>(&self, store: &'a impl AsStoreRef) -> MemoryView<'a> {
```
[the remark this answers is not in the document]
Comment (syrusakbary) [reply lib/api/src/sys/externals/memory.rs:81]: After offline discussion we are keeping it as ref

File: examples/exports_function.rs (hunk 1 above)
Comment (john-sharratt) [reply examples/exports_function.rs]: done

File: examples/errors.rs (hunk 2 above)
Comment (john-sharratt) [reply examples/errors.rs]: done

File: examples/engine_headless.rs (hunk 3 above)
Comment (john-sharratt) [reply examples/engine_headless.rs]: done

File: examples/exports_memory.rs (hunk 4 above)
Comment (john-sharratt) [reply examples/exports_memory.rs]: done

File: examples/features.rs (hunk 5 above)
Comment (john-sharratt) [reply examples/features.rs]: done

File: examples/instance.rs (hunk 6 above)
Comment (john-sharratt) [reply examples/instance.rs]: done

File: examples/memory.rs (hunk 7 above)
Comment (john-sharratt) [reply examples/memory.rs]: done

File: examples/metering.rs (hunk 8 above)
Comment (john-sharratt) [reply examples/metering.rs]: done

File: examples/tunables_limit_memory.rs (hunk 9 above)
Comment (john-sharratt) [reply examples/tunables_limit_memory.rs]: done

File: lib/api/tests/js_externals.rs (hunk 10 above)
Comment (john-sharratt) [reply lib/api/tests/js_externals.rs]: done

File: lib/api/tests/js_externals.rs (hunk 11 above)
Comment (john-sharratt) [reply lib/api/tests/js_externals.rs]: done

File: lib/api/tests/js_externals.rs (hunk 12 above)
Comment (john-sharratt) [reply lib/api/tests/js_externals.rs]: done

File: lib/api/src/js/externals/memory_view.rs (hunk 13 above)
Comment (john-sharratt) [reply lib/api/src/js/externals/memory_view.rs]: done

File: lib/api/Cargo.toml (hunk 14 above)
Comment (john-sharratt) [reply lib/api/Cargo.toml]: done

File: lib/api/tests/js_instance.rs (hunk 15 above)
Comment (john-sharratt) [reply lib/api/tests/js_instance.rs]: done

File: lib/api/tests/js_instance.rs (hunk 16 above)
Comment (john-sharratt) [reply lib/api/tests/js_instance.rs]: done

File: lib/api/tests/js_instance.rs [hunk 19]
```diff
@@ -535,12 +535,12 @@ mod js {
         let instance = Instance::new(&mut store, &module, &import_object).unwrap();
 
         let memory = instance.exports.get_memory("memory").unwrap();
-        assert_eq!(memory.data_size(&store), 65536);
-        let memory_val = memory.uint8view(&store).get_index(0);
+        assert_eq!(memory.view(&store).data_size(), 65536);
+        let memory_val = memory.view(&store).uint8view().get_index(0);
```
[the remark this answers is not in the document]
Comment (john-sharratt) [reply lib/api/tests/js_instance.rs]: done

File: lib/api/tests/js_instance.rs (hunk 17 above)
Comment (john-sharratt) [reply lib/api/tests/js_instance.rs]: done

**Comment by john-sharratt** [comment]
@syrusakbary Please don't squash the two commits, I have another branch I don't want to rebase off this one (too many merge conflicts!)

**Review by syrusakbary — APPROVED** [review]

**Pull request merged**